crates/core/component/dex/src/component/action_handler/position/withdraw.rs

```diff
@@ -27,25 +22,22 @@ impl ActionHandler for PositionWithdraw {
         // we need to ensure that we're checking the reserves at the moment we execute
         // the withdrawal, to prevent any possibility of TOCTOU attacks.
 
-        let mut metadata = state
-            .position_by_id(&self.position_id)
-            .await?
-            .ok_or_else(|| anyhow!("withdrew from unknown position {}", self.position_id))?;
+        // Execute the withdrawal, extracting the reserves from the position.
+        let actual_reserves = state
+            .withdraw_position(self.position_id, self.sequence)
+            .await?;
 
-        // First, check that the commitment to the amount the user is
+        // Next, and CRITICALLY, check that the commitment to the amount the user is
         // withdrawing is correct.
         //
         // Unlike other actions, where a balance commitment is used for
         // shielding a value, this commitment is used for compression, giving a
         // single commitment rather than a list of token amounts.
-
+        //
         // Note: since this is forming a commitment only to the reserves,
         // we are implicitly setting the reward amount to 0. However, we can
         // add support for rewards in the future without client changes.
-        let expected_reserves_commitment = metadata
-            .reserves
-            .balance(&metadata.phi.pair)
-            .commit(Fr::zero());
+        let expected_reserves_commitment = actual_reserves.commit(Fr::zero());
```

Comment:
ACK: I am convinced that this is correct because `position_withdraw` returns the withdrawn reserves